### PR TITLE
Basic translation functionality

### LINKED ISSUE #40: Language translations
From Eileen:
> A major barrier to access right now is English language. We're working on sourcing some help with figuring out whether Google Translate is accurate enough to be useful and for sure at least translating the column headings, legend, and other things that are more or less static, but if someone on the dev side could start thinking sooner rather than later about how to change the interface to make language switching easy and obvious, that would be great. Languages we're shooting for are currently Spanish, Somali, Hmong, and Karen; open to other suggestions.

### PR DESCRIPTION
**Pull request by andyinabox**
This incorporates some very basic, targeted translation which can be edited in this Google Sheet: https://docs.google.com/spreadsheets/d/1m5QPNP6O8nsQsqJcQbwib_obpXRjmbYYGtwKzzg1l38/edit?pli=1#gid=0 (slack me and I'll give you edit access)

It's pretty rough, just trying to get something in there for now. Feel free to add comments/changes since this is really just a rough draft.

Resolves #40

### CODE REVIEW

**Comment by andyinabox** [comment]
@hinzed1127 I didn't include your suggestion for auto language detection yet, figured that could be a next step

**Comment by jdalt** [comment]
@andyinabox Can you elaborate on the strategy you're taking here and what changes will be necessary in the spreadsheet to accommodate them? How might this affect QA in the future?

File: index.html [hunk 1]
```diff
@@ -32,7 +32,8 @@
 	<script>
 		mapboxgl.accessToken = 'pk.eyJ1IjoiamFjb2JyZGFsdG9uIiwiYSI6ImNrYXc2anFjbjIxNGwyeG14dWk0MDVycmIifQ.yILWplracVMB4mv1dWwTtg'
 		const DATA_URL = 'https://spreadsheets.google.com/feeds/list/1CyPozeKhmOuIaVnKDQAUKKsvA9R4F20hEQ3MSGWczP8/1/public/full?alt=json'
-	</script>
+		const TRANS_URL = 'https://spreadsheets.google.com/feeds/list/1m5QPNP6O8nsQsqJcQbwib_obpXRjmbYYGtwKzzg1l38/1/public/full?alt=json'
```
Comment (jdalt) [inline index.html:35]: Is this basically a copy of the original spreadsheet that is translated whole hog? Are just parts translated?

File: index.html [hunk 2]
```diff
@@ -41,10 +42,17 @@
...
   <div class="content">
     <div class="title">
-      <h1 class='h1'>Twin Cities Aid Map</h1>
+      <h1 data-translation-id="page_title" class='h1'>Twin Cities Aid Map</h1>
+      <ul class="languages">
+        <li><a data-lang="eng" href="/">English</a></li>
+        <li><a data-lang="spa" href="/?lang=spa">Español</a></li>
+        <li><a data-lang="kar" href="/?lang=kar">Karen</a></li>
+        <li><a data-lang="som" href="/?lang=som">Soomaali</a></li>
+        <li><a data-lang="hmn" href="/?lang=hmn">Hmong</a></li>
```
Comment (jdalt) [inline index.html:51]: It sounds like we might have folks that can do Amharic and Oromo too. Doesn't have to be added here. Can be a followup - just want you to know.

**Comment by kmarekspartz** [comment]
Clicking Español didn't change the text for me, on Mac Safari.

File: translator.js [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+  /**
+   * Replace the translatable elements on the page with translated terms (if available)
+   * 
+   * @param {string} lang key for the desired language (example: 'spa')
+   */
+  translate(lang) {
+    this.els = document.querySelectorAll(`[${this.attr}]`) 
+    
+    let current
+    for (let i = 0; i < this.els.length; i++) {
```
Comment (jdalt) [inline translator.js:36]: Is there any reason this can't be done in the functional `els.forEach(` style?

**Comment by kmarekspartz** [comment]
Syntax errors, looks similar to #28 

[Error] SyntaxError: Unexpected token '='. Expected an opening '(' before a method's parameter list.
	(anonymous function) (translator.js:16)
[Error] Error translating page – ReferenceError: Can't find variable: Translator
ReferenceError: Can't find variable: Translator
	(anonymous function) (script.js:279)
	asyncFunctionResume
	(anonymous function)
	promiseReactionJobWithoutPromise

File: translator.js [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+class Translator {
+  attr = 'data-translation-id'
+
+  constructor(translations) {
+    this.translations = translations
+  }
+
+  // this converts a `gsx$keyname` to `keyname
+  convertGoogleKey(key) {
```
Comment (jdalt) [inline translator.js:23]: Is this just stubbed out or is it overridden somewhere? It's not clear to me why this is returning null or undefined.

**Comment by andyinabox** [comment]
@jdalt no changes are needed to the spreadsheet, I am only translating labels here ... not data from the spreadsheet (that is a much bigger job). I left some notes about implementation here: https://github.com/Twin-Cities-Mutual-Aid/twin-cities-aid-distribution-locations/issues/40

@kmarekspartz thanks, hadn't tested in Safari yet!

**Comment by cwood821** [comment]
This is awesome!

Someone could disagree with me here, one tweak might be to make more of the Spanish filters noun phrases. For example, “Necesita dinero” > “necesidad de dinero”, “necessita voluntarias” > “necesidad de voluntarios”. That would line up with some of the existing translations like “estado de ubicacion”.

**Comment by andyinabox** [comment]
@cwood821 I think we will probably have another process for vetting the actual translations ... I just used Google translate so I could show an example

File: index.html (hunk 1 above)
Comment (experimatt) [reply index.html:35]: This is a nitpick but I think using `TRANSLATION_URL` for the variable name, while longer, would be more appropriate.

File: index.html (hunk 2 above)
Comment (andyinabox) [reply index.html:51]: @experimatt I am using https://en.wikipedia.org/wiki/List_of_ISO_639-2_codes verses the more common ones like `en` or `en-US` to support languages like Hmong and Karen. I wasn't able to find standard 2-letter language codes for those.

File: index.html (hunk 1 above)
Comment (andyinabox) [reply index.html:35]: @jdalt only static labels like column headers, map legend, etc are being translated. Translating the actual data is a huge lift but this gets us a step closer

File: translator.js (hunk 3 above)
Comment (andyinabox) [reply translator.js:36]: `querySelectorAll` returns a `NodeList` (not an array) which doesn't implement `forEach.` I'll convert to an array instead and use `forEach` for code consistency

File: index.html (hunk 2 above)
Comment (experimatt) [reply index.html:51]: @andyinabox ahhhh gotcha, I didn't realize that. carry on then :)

**Comment by andyinabox** [comment]
Push fixes/adjustments based on comments above. Some highlights:

 * Hoping I fixed the Safari issue @kmarekspartz mentioned, although cant test currently. Can anyone verify?
 * Switched to use a `<select>` for languages as @experimatt suggestion. I think it works ok, but my one hesitation is that it might be less visible. Before you could see "Soomaali" right on the page, but now you have to know to click the language dropdown first

A few additional notes

 * Didn't add any flag icons, partially because some of the languages (e.g. Hmong) don't have a corresponding flag
 * The legend/title area is getting pretty bloated but I think that would best be handled in a separate UI task
 * Currently disabled all languages except English and Spanish since we don't have translations yet for the rest (maybe Spanish shouldn't even be enabled since it's just from Google translate)

File: index.html (hunk 2 above)
Comment (jdalt) [reply index.html]: til another iso code

File: translator.js (hunk 3 above)
Comment (jdalt) [reply translator.js]: silly dom api

**Comment by jdalt** [comment]
Safari looks good.

**Comment by kmarekspartz** [comment]
Confirmed, works on Safari.

I'm not sold on the dropdown... If I'm not an English speaker, how would I know to click on the Language: English button to find e.g. Hmoob?

**Comment by jdalt** [comment]
I really want to merge this but the mobile experience is hurting right now. Here's the chrome simulated iphone 5/6se view:
Twin_Cities_Aid_Distribution_Locations

It's sort of unfair because you didn't create this problem but it's definitely worse with another vertical entry.

Let's talk in the morning about ways that we can improve the mobile experience.

**Comment by kmarekspartz** [comment]
Other thought: feature toggle url parameter for testing while the translations get built out?

–
Kyle Marek-Spartz

**Comment by andyinabox** [comment]
Saman had a good suggestion in #57 

> Consider handling translation with an immediate popup the first time user visits the site that asks user for preferred language, and then saving the result to localstorage. We can then put the language toggle behind help/info.

**Comment by andyinabox** [comment]
Update on this: it's working, although I haven't tested extensively or really been able to do a once-over to clean up the code. 

Main new feature here is a "welcome" modal that shows up if the user hasn't visited before.

I have to go back to work, but wanted to get this up there for review. Things that I'm aware of that still need to be done:

1. Refine styling for welcome modal (this could be done by somebody else)
2. Fix merge conflicts (probably needs to be me)
3. Move translation drop-down to "info" section (probably me, after fixing merge conflicts)
4. QA (can be done by somebody else)
5. Add translations (can be somebody else)

**Comment by experimatt** [comment]
I can take a stab at resolving the merge conflicts, seeing as I just added some code that re-organized a few things

**Comment by experimatt** [comment]
Merge conflicts resolved. We're going to need to add some additional rows to the translation spreadsheet to reflect additional content on the site that's been added recently.

File: welcome.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    })
+  }
+
+  render(languages) {
+
+    const buttons = languages.map(l => `<button class="language-button" value="${l.name}">${l.label}</button>`).join('')
+
+    return `
+      <div class="modal">
+        <h1 class="welcome-message">Welcome!</h1>
```
Comment (samanpwbb) [inline welcome.js:29]: A snappy one-liner/purpose statement would be useful here I think:

```suggestion
        <h1 class="welcome-message">Welcome! Twin Cities Mutual Aid is an up to date resource on mutual aid sites in the Twin Cities.</h1>
```

File: welcome.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+      })
+    })
+  }
+
+  render(languages) {
+
+    const buttons = languages.map(l => `<button class="language-button" value="${l.name}">${l.label}</button>`).join('')
+
+    return `
+      <div class="modal">
```
Comment (samanpwbb) [inline welcome.js:28]: Would it be possible to make it clear that this is a modal / hint at the content behind the modal by adding margin around the edges & not making the modal 100% height?

File: welcome.js (hunk 5 above)
Comment (experimatt) [reply welcome.js]: @samanpwbb good call, I just added this!

File: welcome.js (hunk 6 above)
Comment (experimatt) [reply welcome.js:28]: Good idea

**Comment by benjifs** [comment]
Related to this PR but could be left for later:

We could also replace the labels in the popups from the map by adding `data-translation-id='${key}'` and adding those translations from the map.

https://github.com/Twin-Cities-Mutual-Aid/twin-cities-aid-distribution-locations/blob/a41acaf23bed16fde0745ce8c16350167dc58201/script.js#L260-L264

Since its currently using the `camelToTitle` function, it would just be adding those keys to the table and translating them. We would need to trigger `translate` on popup appearing as well which is why maybe we can just leave this for another issue.

Unrelated but the keys there have some spelling issues so a quick typo fix there would be great.

https://github.com/Twin-Cities-Mutual-Aid/twin-cities-aid-distribution-locations/blob/a41acaf23bed16fde0745ce8c16350167dc58201/script.js#L233-L239

File: welcome.js (hunk 6 above)
Comment (andyinabox) [reply welcome.js:28]: Yeah good idea, thanks!

**Comment by andyinabox** [comment]
Ok, one more round on this one.

 * Preview with only translated languages visible: https://deploy-preview-54--wonderful-montalcini-96d80d.netlify.app/
 * Preview with all languages visible: https://deploy-preview-54--wonderful-montalcini-96d80d.netlify.app/?dev

The UX has change significantly, how y'all are onboard for the updates.

1. Still doing the "Welcome" modal when user first visits
2. Language is displayed in the top left corner of the map
3. Click on the language will re-open the "Welcome" modal

thanks to @experimatt for diving in on this when I wasn't available!

**Comment by andyinabox** [comment]
@benjifs yeah, I've had that in the back of my mind but was trying to keep things more limited to get out the door. But probably makes sense to add all the headers from the main data source to the translation doc, so we don't have to go back to the translators more than once

**Comment by benjifs** [comment]
@andyinabox just added the keys and English and Spanish translations for those. I'm also adding some changes in a different branch which would add the translations to the map popups

**Comment by samanpwbb** [comment]
I can do a really quick design cleanup/consistification pass here in about 20 mins.

**Comment by andyinabox** [comment]
@benjifs Awesome! Thanks for doing that. Couldn't sleep, so I'm gonna make the change you suggested :grin:. Will probably switch the ids to snake_case for consistency

**Comment by samanpwbb** [comment]
I hit this error every time when running from localhost and at the staging URL – is this a known issue?

Screen Shot 2020-06-04 at 9 55 55 PM

**Comment by samanpwbb** [comment]
Just pushed a really quick design cleanup (added max width to modal, cleaned up spacing):

Screen Shot 2020-06-04 at 9 58 21 PM

And on desktop:
Screen Shot 2020-06-04 at 9 58 13 PM

File: translator.js [hunk 7]
```diff
@@ -40,35 +49,44 @@ class Translator {
...
     window.localStorage.setItem(Translator.LOCAL_STORAGE_KEY, lang)
-
   }
 
+  /**
+   * Get current language
+   */
   get language() {
     const lang = window.localStorage.getItem(Translator.LOCAL_STORAGE_KEY)
-    if (this.validateLanguageKey(lang)) return lang
```
Comment (benjifs) [inline translator.js]: @samanpwbb This line is causing the issue. If I add it back it works for me

**Comment by andyinabox** [comment]
@samanpwbb oh yeah sorry, I did a cleanup pass and was a little overzealous! Added a comment explaining why that line is necessary.

Just pushed @benjifs's suggestion for setting the map popup labels. Also fixed issue where sidebar cards weren't being translated on filter change.

**Comment by andyinabox** [comment]
also those design tweaks look great!

**Comment by experimatt** [comment]
This looks so great, amazing work everyone!

File: script.js [hunk 8]
```diff
@@ -160,7 +210,9 @@ const createListItem = (location, status, lng, lat) => {
         essential: true,
         zoom: 13
       })
+      console.log('popup', popup)
```
Comment (experimatt) [inline script.js:213]: errant console I wonder if we can remove this

Comment (experimatt) [reply script.js:213]: Couldn't get this console.log to run locally, so I left it in. :yolo:

**Comment by experimatt** [comment]
I just tested this locally and in netlify and think it looks great. merging!

**Comment by experimatt** [comment]
We reverted this, then unreverted it once we cleaned up the underlying data issues. whoops

**Pull request merged**